### PR TITLE
ARROW-3608: [R] Support for time32 and time64 array types

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -298,6 +298,8 @@ matrix:
     - bash <(curl -s https://codecov.io/bash) || echo "Codecov did not collect coverage reports"
   # R
   - language: r
+    env:
+    - GITHUB_PAT="4aa63e8f11dbebae5a51018c3e8d5bfa269947f7"
```
Comment (wesm) [inline .travis.yml]: Is this safe?

File: r/src/array.cpp [hunk 2]
```diff
@@ -373,6 +373,75 @@ std::shared_ptr<arrow::Array> Int64Array(SEXP x) {
...
+}
+
+std::shared_ptr<arrow::Array> Time32Array_From_difftime(SEXP x) {
+  // number of seconds as a double
+  auto p_vec_start = REAL(x);
+  auto n = Rf_xlength(x);
+  int64_t null_count = 0;
+
+  int multiplier = difftime_unit_multiplier(x);
+  std::vector<std::shared_ptr<Buffer>> buffers{nullptr, nullptr};
```
Comment (wesm) [inline r/src/array.cpp]: `buffers(2)` works too

File: r/src/array.cpp [hunk 3]
```diff
@@ -373,6 +373,75 @@ std::shared_ptr<arrow::Array> Int64Array(SEXP x) {
...
+  R_xlen_t i = 0;
+  auto p_vec = p_vec_start;
+  for (; i < n; i++, ++p_vec, ++p_values) {
+    if (NumericVector::is_na(*p_vec)) {
+      break;
+    }
+    *p_values = static_cast<int32_t>(*p_vec * multiplier);
+  }
+
+  if (i < n) {
```
Comment (wesm) [inline r/src/array.cpp]: I have seen this pattern in some other conversion functions. Seems like you could introduce some lambdas / functors and have a reusable template for this part

File: r/src/array.cpp [hunk 4]
```diff
@@ -698,6 +770,34 @@ SEXP Int64Array(const std::shared_ptr<Array>& array) {
...
+    arrow::internal::BitmapReader bitmap_reader(array->null_bitmap()->data(),
+                                                array->offset(), n);
+    for (size_t i = 0; i < n; i++, bitmap_reader.Next(), ++p_vec, ++p_values) {
+      *p_vec =
+          bitmap_reader.IsSet() ? (static_cast<double>(*p_values) / multiplier) : NA_REAL;
+    }
+  } else {
+    std::transform(p_values, p_values + n, vec.begin(), [multiplier](value_type value) {
+      return static_cast<double>(value) / multiplier;
+    });
```
Comment (wesm) [inline r/src/array.cpp]: Same here, could pass a lambda into an inline template

**Review by wesm — COMMENTED** [review]
This looks OK, couple implementation suggestions and wasn't sure whether you wanted to check in this PAT

File: r/src/array.cpp (hunk 3 above)
Comment (romainfrancois) [reply r/src/array.cpp]: Yes, there is some repetition. I'll write something more generic, but I can do that in a follow up.

**Pull request closed without merging**